pageserver/src/page_service.rs

```diff
@@ -346,21 +346,26 @@ impl PageServerHandler {
         pgb.write_message(&BeMessage::CopyOutResponse)?;
         info!("sent CopyOut");
 
-        /* Send a tarball of the latest snapshot on the timeline */
-
-        let req_lsn = match lsn {
-            Some(lsn) => {
-                timeline.wait_lsn(lsn)?;
-                lsn
-            }
-            None => timeline.get_last_valid_lsn(),
+        // Currently we use empty lsn in two cases:
+        //  * During basebackup right after timeline creation
+        //  * When working without safekeepers. In this situation it is important to match the lsn
+        //    we are taking basebackup on with the lsn that is used in pageserver's walreceiver
+        //    to start the replication. E.g. if we are waiting here for tha last_valid_lsn compute
+        //    node after restart will not necessarily have last_record_lsn.
+        let backup_lsn = if let Some(req_lsn) = lsn {
+            timeline.wait_lsn(req_lsn)?;
+            req_lsn
+        } else {
+            timeline.get_last_record_lsn()
         };
+
+        /* Send a tarball of the latest snapshot on the timeline */
         {
             let mut writer = CopyDataSink { pgb };
             let mut basebackup = basebackup::Basebackup::new(
                 &mut writer,
                 &timeline,
-                req_lsn,
+                backup_lsn,
                 timeline.get_prev_record_lsn(),
```

Comment:
+1 to that and follow up comments, will fix

Reply:
Hm, seems that problem is bigger and not directly related to this patch: https://github.com/zenithdb/zenith/issues/506

It is straightforward to fix race but we do not have prev_record_lsn for older lsn. Not yet sure what to do, but probably that is another PR since behavior exists in main.